### PR TITLE
Fix TextEditorPresenter spec timeout.

### LINKED ISSUE #10435: Flaky text-editor-presenter spec
https://travis-ci.org/atom/atom/jobs/102501079

Internal CI passed for this commit.

/cc @kuychaco

### PR DESCRIPTION
**Pull request by joshaber**
Fixes #10435.

I haven’t been able to reproduce this locally yet, but let’s see if it’s as easy as giving it a longer timeout.

### CODE REVIEW

**Comment by joshaber** [comment]
A parallel `atom` build failed: https://janky.githubapp.com/2759555/output

This kinda seems like a legit race.

**Comment by joshaber** [comment]
This is ready.

The problem was with `waitsForStateToUpdate`. We'd call a function that emitted a state update, and _then_ start watching for state updates. So we were racing between emitting the event and watching for it.

I introduced `waitsForStateToUpdateAsync` instead of fixing `waitsForStateToUpdate`, because when I changed `waitsForStateToUpdate` it broke some other tests which I didn't understand.

/cc @as-cii @nathansobo :eyes: I think you two did most of the work here.

File: spec/text-editor-presenter-spec.coffee [hunk 1]
```diff
@@ -20,6 +20,7 @@ describe "TextEditorPresenter", ->
 
       buffer = new TextBuffer(filePath: require.resolve('./fixtures/sample.js'))
       editor = atom.workspace.buildTextEditor({buffer})
+      editor.setUpdatedSynchronously(true)
```
Comment (nathansobo) [inline spec/text-editor-presenter-spec.coffee]: Interested in why this was needed...

**Comment by nathansobo** [comment]
Thanks for this. Would be nice to only have one `waitsForStateToUpdate` method, but I'd rather fix the flakiness and get back to it later I suppose.

File: spec/text-editor-presenter-spec.coffee (hunk 1 above)
Comment (joshaber) [inline spec/text-editor-presenter-spec.coffee]: It wasn't needed, but it made the test failures deterministic so I could ensure it was fixed.

Comment (nathansobo) [inline spec/text-editor-presenter-spec.coffee]: Ah, do you think we could drop it now then? I feel like the test is more realistic overall without it.

**Comment by joshaber** [comment]
>Would be nice to only have one waitsForStateToUpdate method, but I'd rather fix the flakiness and get back to it later I suppose.

Yeah, I agree. I time boxed this so I could get to some of the more pressing FR issues. Maybe the next FR could dig into this further.

File: spec/text-editor-presenter-spec.coffee (hunk 1 above)
Comment (joshaber) [inline spec/text-editor-presenter-spec.coffee]: :+1:

**Comment by nathansobo** [comment]
> Yeah, I agree. I time boxed this so I could get to some of the more pressing FR issues. Maybe the next FR could dig into this further.

Totally reasonable.

**Comment by joshaber** [comment]
We're back down to one `waitsForStateToUpdate` now :metal:

**Comment by nathansobo** [comment]
> We're back down to one waitsForStateToUpdate now :metal:

:pray: Namaste

**Comment by maxbrunsfeld** [comment]
:moneybag: :train: :dollar: :steam_locomotive:!!

**Comment by as-cii** [comment]
Thank you, @joshaber! :boom:

**Pull request merged**